cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+
+      const double diffusion_viscosity = std::pow(diffusion_creep_prefactor[phase_index],-1.0/diffusion_creep_exponent[phase_index])
+                                         * std::pow(second_strain_rate_invariant,strain_rate_dependence)
+                                         * std::pow(grain_size, diffusion_creep_grain_size_exponent[phase_index]/diffusion_creep_exponent[phase_index])
+                                         * energy_term;
+
+      Assert(diffusion_viscosity > 0.0,
+             ExcMessage("Diffusion viscosity has to be positive. Instead it is: " + std::to_string(diffusion_viscosity)));
+
+      return diffusion_viscosity;
+    }
+
+
+
+    template <int dim>
+    double
+    EquilibriumGrainSize<dim>::
+    dislocation_viscosity (const double      temperature,
+                           const double      pressure,
+                           const std::vector<double> &composition,
+                           const SymmetricTensor<2,dim> &strain_rate,
+                           const Point<dim> &position,
+                           const double viscosity_guess) const
+    {
+      const double diff_viscosity = diffusion_viscosity(temperature,pressure,composition,strain_rate,position);
+
+      // If we do not have the strain rate yet, set it to a low value.
+      // TODO: make minimum strain rate an input parameter
+      const SymmetricTensor<2,dim> shear_strain_rate = strain_rate - 1./dim * trace(strain_rate) * unit_symmetric_tensor<dim>();
+      const double second_strain_rate_invariant = std::max(std::sqrt(std::abs(second_invariant(shear_strain_rate))),1e-30);
+
+      // Start the iteration with the full strain rate
+      double dis_viscosity = viscosity_guess;
+      if (viscosity_guess == 0)
+        dis_viscosity = dislocation_viscosity_fixed_strain_rate(temperature,pressure,second_strain_rate_invariant,position);
+
+      double dis_viscosity_old = 0;
+      unsigned int i = 0;
+      while ((std::abs((dis_viscosity-dis_viscosity_old) / dis_viscosity) > dislocation_viscosity_iteration_threshold)
+             && (i < dislocation_viscosity_iteration_number))
+        {
+          const double dislocation_strain_rate = diff_viscosity / (diff_viscosity + dis_viscosity) * second_strain_rate_invariant;
+          dis_viscosity_old = dis_viscosity;
+          dis_viscosity = dislocation_viscosity_fixed_strain_rate(temperature,
+                                                                  pressure,
+                                                                  dislocation_strain_rate,
+                                                                  position);
+          ++i;
+
+          Assert(dis_viscosity > 0.0,
+                 ExcMessage("Encountered negative dislocation viscosity in iteration " + std::to_string(i) +
+                            ". Dislocation viscosity is: " + std::to_string(dis_viscosity)));
```

Comment:
You first increment `i` and then use it in the generation of the error message. Is that on purpose? Wouldn't you want the error message to reflect the `i` that was used to compute the quantity you are complaining about in the error message?

Reply:
I guess it is because our value of i is 0 for the first iteration and we would want to say "iteration 1" in our message.

Reply:
I think that makes sense, but I would also be fine with switching it.